oak_tdx_guest/src/vmcall.rs

```diff
@@ -0,0 +1,528 @@
+/// for more information.
+fn io_write(port: u32, size: IoWidth, data: u64) -> Result<(), &'static str> {
+    // The VMCALL sub-function for Instruction.IO.
+    const SUB_FUNCTION: u64 = 30;
+    // The result if the IO operation was invalid.
+    const INVALID_OPERAND: u64 = 1 << 63;
+
+    let mut vm_call_result: u64;
+    let mut sub_function_result: u64;
+    let registers = Registers::default()
+        .union(Registers::R12)
+        .union(Registers::R13)
+        .union(Registers::R14)
+        .union(Registers::R15);
+
+    // The TDCALL leaf 0 goes into RAX. RAX returns the top-level result (0 is success). The
+    // bitflags of registers to be passed through to the VMM goes into RCX. The sub-function
+    // usage (always 0 when conforming to the GHCI spec) goes into R10, and the result of the
+    // subfunction is returned in R10. The sub-function to call goes into R11. The size of the
+    // write (1,2 or 4 bytes) goest into R12. The direction (write) goes into R13. The IO port
+    // number goes into R14. The data to write goes into R15.
+    //
+    // Safety: calling TDCALL here is safe since it does not alter memory and all the affected
+    // registers are specified, so no unspecified registers will be clobbered.
+    unsafe {
+        asm!(
+            "tdcall",
+            inout("rax") VM_CALL_LEAF => vm_call_result,
+            in("rcx") registers.bits,
+            inout("r10") DEFAULT_SUB_FUNCTION_USAGE => sub_function_result,
+            in("r11") SUB_FUNCTION,
+            in("r12") size as u64,
+            in("r13") IoDirection::Write as u64,
+            in("r14") port as u64,
+            in("r15") data,
+            options(nomem, nostack),
+        );
+    }
+
+    // According to the spec the top-level result for this sub-function will aways be 0 as long as
+    // the specified sub-function leaf is correct.
+    assert_eq!(
+        vm_call_result, SUCCESS,
+        "TDG.VP.VMCALL returned an invalid result"
+    );
+
+    if sub_function_result == INVALID_OPERAND {
+        return Err("IO write operation failed");
+    }
+    // According to the spec this sub-function will always return either 0 or INVALID_OPERAND.
+    assert_eq!(
+        sub_function_result, SUCCESS,
+        "TDG.VP.VMCALL<Instruction.IO> returned an invalid result"
+    );
+
+    Ok(())
+}
+
+/// Writes a value to the specified model-specific register.
+///
+/// See section 3.11 of [Guest-Host-Communication Interface (GHCI) for Intel® Trust Domain
+/// Extensions (Intel® TDX)](https://www.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-guest-hypervisor-communication-interface.pdf)
+/// for more information.
+///
+/// #Safety
+///
+/// Modifying an MSR could impact the execution environment, so the caller must ensure that the MSR
+/// modification will not lead to undefined behavior.
+pub unsafe fn msr_write(msr: u32, data: u64) -> Result<(), &'static str> {
```

Comment:
In general, I wonder if we should enable this: https://doc.rust-lang.org/beta/nightly-rustc/rustc_lint/builtin/static.UNSAFE_OP_IN_UNSAFE_FN.html

I find it confusing that `unsafe fn` implies that the entire body is itself an `unsafe { }` block.

WDYT?

Reply:
I think that is a good idea. I added it to this crate and create https://github.com/project-oak/oak/issues/3920 to track it more widely.